tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/LazyEquality.scala

```diff
@@ -424,10 +424,17 @@ class LazyEquality(rewriter: SymbStateRewriter)
       case `arraysEncoding` =>
         // In the arrays encoding we only cache the equalities between the elements of the functions' ranges
         // This is because the ranges consist of pairs of form <arg,res>, thus the domains are also handled
+        // TODO: stop caching unnecessary equalities, similar to SetInRuleWithArrays.basicIn?
         val leftElems = state.arena.getHas(leftRel)
         val rightElems = state.arena.getHas(rightRel)
         cacheEqConstraints(state, leftElems.cross(rightElems)) // Cache all the equalities
         eqCache.put(leftFun, rightFun, EqCache.EqEntry())
+        // For the rare case in which one function has an empty domain, we need to be extra careful
+        // See https://github.com/informalsystems/apalache/issues/1811
+        val leftDom = state.arena.getDom(leftFun)
+        val rightDom = state.arena.getDom(rightFun)
+        rewriter.solverContext.assertGroundExpr(tla.eql(leftDom.toNameEx, rightDom.toNameEx))
```

Comment:
I don't think this is sound, as this constraint unconditionally requires the two domains to be equal. This condition should be a part of the function equality.

Reply:
Good point. `LazyEquality` don't make it easy to add constraints to the cached equalities, so I think the best way to go about it is to assert `funEq => domEq`.